### PR TITLE
Add support for absolute paths in DataFrameIterator

### PR DESCRIPTION
**Pull request by smurak**
### Summary

Add support for absolute paths in DataFrameIterator.

### Related Issues

#56, #67

### PR Overview

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [x] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

File: keras_preprocessing/image.py [hunk 1]
```diff
@@ -2098,13 +2101,21 @@ def __init__(self, dataframe, directory, image_data_generator,
...
+                white_list_formats,
+                self.split,
+                class_indices=self.class_indices,
+                follow_links=follow_links,
+                df=True)
+        else:
+            if not has_ext:
+                raise ValueError('has_ext cannot be set to False'
+                                 ' if directory is None.')
+            filenames = list(self.df[x_col][self.df[x_col].map(os.path.exists)])
```
Comment (Vijayabhaskar96) [inline keras_preprocessing/image.py]: `os.path.exists` will return turn true for both directories and files, `os.path.isfile` will return true only for files,So I think it's better to use  `os.path.isfile` here.

**Comment by smurak** [comment]
@Vijayabhaskar96 Thank you, I changed some codes.

- Change `os.path.exists` to `os.path.isfile`.
- Add a file format check and tests.
- Move the "split" part to the last part.
  (do not split `filenames` but split `self.filenames`.)

**Comment by Vijayabhaskar96** [comment]
@smurak Thanks for PR. @Dref360 Looks good to me.

**Comment by smurak** [comment]
@Vijayabhaskar96 Thank you for reviewing.
I also added support for capital file extensions (#88), so would you check this again?

**Comment by Vijayabhaskar96** [comment]
@smurak Looks fine.

**Comment by smurak** [comment]
@Vijayabhaskar96 Thank you!

**Comment by Mahi-Mai** [comment]
@smurak   Is there a way for me to checkout and use the branch that accepts capital file extensions, or has that not been committed yet?

**Comment by smurak** [comment]
@Mahi-Mai This PR has been merged to master, 
so you can download this repository and use the fixed version of flow_from_dataframe.

**Comment by egridavid** [comment]
Hi @smurak . For me it is not clear how can I use this new  version of flow_from_directory ? I dowloaded this repository as mentioned but I dont know how can I use it. I am using anaconda on windows.

**Comment by Mahi-Mai** [comment]
@egridavid  Hi!  I was similarly lost.  Here's what I did, but I'm not sure if it's the "right" way to do it, and I'm not sure if it's going to cause issues down the line:

1. Download/clone both the base keras repo and the keras-preprocessing repo. 
2. Replace the contents of the base keras repo's preprocessing folder with the contents of ~./keras-preprocessing/keras_preprocessing
3.  I changed the name of my altered base keras repo so it looked like /path/keras2/keras2/contents .  This let me specifically choose to load the altered keras instead of what's already installed on my computer by pip.  I load it like this:

```
import sys
sys.path.append('/Users/myname/documents/tools/keras2/')
from keras2.preprocessing.image import ImageDataGenerator
```

I imagine when I make all my other keras imports, I'll need to be careful to reference keras2 instead of keras.  This seems hacky, but it's working!

**Comment by smurak** [comment]
@Mahi-Mai I think your way is ok, but more easy way would be:
1. Download this repository.
2. Move the "keras_preprocessing" subdirectory to your working directory.
3. Import it.
    ```
    import keras
    from keras_preprocessing.image import ImageDateGenerator
    ```

**Comment by Mahi-Mai** [comment]
I ask all this because I'm getting some odd results from my model training and want to make sure my method outlined above isn't breaking something under the hood.

@smurak Your doesn't appear to work.  My code:
```
import pandas as pd
import numpy as np

import keras
import sys
sys.path.append('/Users/name/documents/tools/keras-preprocessing/keras_preprocessing')
from keras_preprocessing.image import ImageDataGenerator

train = pd.read_csv('absolute directories.csv', delimiter='\t')

main_dir = None

img_width, img_height = 224, 224
batch_size = 20

train_datagen = ImageDataGenerator(horizontal_flip = True,
                                   fill_mode = "nearest",
                                   zoom_range = 0.3,
                                   width_shift_range = 0.1,
                                   height_shift_range = 0.1,
                                   rotation_range = 30)

train_generator = train_datagen.flow_from_dataframe(dataframe=train,
                                                    directory=main_dir,
                                                    x_col='location',
                                                    y_col='label',
                                                    has_ext=True,
                                                    target_size = (img_height, img_width),
                                                    batch_size = batch_size, 
                                                    class_mode = "binary")
```
Get's me this error:
```
TypeError                                 Traceback (most recent call last)
<ipython-input-4-5e05e28f635d> in <module>()
     22                                                     target_size = (img_height, img_width),
     23                                                     batch_size = batch_size,
---> 24                                                     class_mode = "binary")

/anaconda3/lib/python3.6/site-packages/keras_preprocessing/image.py in flow_from_dataframe(self, dataframe, directory, x_col, y_col, has_ext, target_size, color_mode, classes, class_mode, batch_size, shuffle, seed, save_to_dir, save_prefix, save_format, subset, interpolation)
   1105                                  save_format=save_format,
   1106                                  subset=subset,
-> 1107                                  interpolation=interpolation)
   1108 
   1109     def standardize(self, x):

/anaconda3/lib/python3.6/site-packages/keras_preprocessing/image.py in __init__(self, dataframe, directory, image_data_generator, x_col, y_col, has_ext, target_size, color_mode, classes, class_mode, batch_size, shuffle, seed, data_format, save_to_dir, save_prefix, save_format, follow_links, subset, interpolation, dtype)
   2093             class_indices=self.class_indices,
   2094             follow_links=follow_links,
-> 2095             df=True)
   2096         if has_ext:
   2097             ext_exist = False

/anaconda3/lib/python3.6/site-packages/keras_preprocessing/image.py in _list_valid_filenames_in_directory(directory, white_list_formats, split, class_indices, follow_links, df)
   1762             `["file1.jpg", "file2.jpg", ...]`).
   1763     """
-> 1764     dirname = os.path.basename(directory)
   1765     if split:
   1766         num_files = len(list(

/anaconda3/lib/python3.6/posixpath.py in basename(p)
    144 def basename(p):
    145     """Returns the final component of a pathname"""
--> 146     p = os.fspath(p)
    147     sep = _get_sep(p)
    148     i = p.rfind(sep) + 1

TypeError: expected str, bytes or os.PathLike object, not NoneType
```

If I instead import my hacky version, like I mentioned:
```
import sys
sys.path.append('/Users/name/documents/tools/hacky_keras/')
from hacky_keras.preprocessing.image import ImageDataGenerator
```

It executes fine.

**Comment by smurak** [comment]
I guess you didn't do this part.

> 2. Move the "keras_preprocessing" subdirectory to your working directory.

The error said:
> /anaconda3/lib/python3.6/site-packages/keras_preprocessing/image.py in flow_from_dataframe(...

so, you used another keras_preprocessing.
If you can't move the directory to your working directory, try `sys.path.insert(0, ...)` instead of `sys.path.append(...)`.

**Comment by smurak** [comment]
@Mahi-Mai Try `sys.path.insert(0, '/Users/name/documents/tools/keras-preprocessing')` (i.e. drop '/keras_preprocessing').

**Comment by Mahi-Mai** [comment]
Ack, that won't work either!  I feel like Jupyter Lab is giving anaconda and the first 'import keras' line priority??

**Comment by egridavid** [comment]
@Mahi-Mai For me the solution was the following (you can use pip inside anaconda, but you have to use the anaconda terminal):
1. In the conda terminal: 
* `pip unistall keras-preprocessing`
* `pip install git+https://github.com/keras-team/keras-preprocessing.git`
2. Use this in the notebook:
* ` from keras_preprocessing.image import ImageDataGenerator`

When I manually changed the directories it did not work for me.

**Comment by smurak** [comment]
> Ack, that won't work either!

@Mahi-Mai Did you try this?
```
import keras
import sys
sys.path.insert(0, '/Users/name/documents/tools/keras-preprocessing')
from keras_preprocessing.image import ImageDataGenerator
```
(maybe "name" should be changed to your name.)

And if you can reinstall keras_preprocessing, it's better to do it as @egridavid said.
(I haven't used anaconda, so I'm not sure the way though.)

**Pull request merged**